### PR TITLE
try pushing again when we get unexpected exception...

### PR DESCRIPTION
**Pull request by martenson**
the test never fails on local but fails randomly on Jenkins

### CODE REVIEW

File: test/shed_functional/base/twilltestcase.py [hunk 1]
```diff
@@ -271,6 +271,22 @@ def commit_and_push( self, repository, hgrepo, options, username, password ):
...
             else:
+                return self.__push_again( hgrepo, url )
+        return True
+
+    def __push_again( hgrepo, url ):
+        try:
+            commands.push( ui.ui(), hgrepo, dest=url )
+        except Abort as a:
+            message = a
+            if 'authorization failed' in message:
```
Comment (davebx) [inline test/shed_functional/base/twilltestcase.py]: This may need to be `str(message)` so that the substring match picks it up.

**Comment by martenson** [comment]
@davebx - good point, I removed that check entirely, it is not needed anymore

File: test/shed_functional/base/twilltestcase.py [hunk 2]
```diff
@@ -261,12 +261,16 @@ def commit_and_push( self, repository, hgrepo, options, username, password ):
...
-                raise
+                return self.__push_again( hgrepo, url )
+        return True
+
+    def __push_again( hgrepo, url ):
+        try:
+            commands.push( ui.ui(), hgrepo, dest=url )
         except Exception as e:
             if str(e).find('Pushing to Tool Shed is disabled') != -1:
                 return False
```
Comment (nsoranzo) [inline test/shed_functional/base/twilltestcase.py]: This is returning `Null` if `commands.push()` succeeds, which is probably not what you want.

Maybe use a `for` cycle in `commit_and_push()`:

```
        for i in range(2):
            try:
                commands.push( ui.ui(), hgrepo, dest=url )
            except Exception as e:
                if str(e).find('Pushing to Tool Shed is disabled') != -1:
                    return False
            else:
                return True
        raise
```

In this way you don't need the `__push_again` method, i.e. less code duplication!

Comment (martenson) [inline test/shed_functional/base/twilltestcase.py]: I think it returns True on L279 if the push succeeds

Comment (nsoranzo) [inline test/shed_functional/base/twilltestcase.py]: @martenson You are right, sorry, I was confused by how GitHub showed the diff. Anyway, I think avoiding code duplication would be good.

Comment (martenson) [inline test/shed_functional/base/twilltestcase.py]: I think this is consistent with how we were treating these Jenkins-only fails in the past and reads nicer in my opinion.

**Comment by martenson** [comment]
ready for review

File: test/shed_functional/base/twilltestcase.py (hunk 2 above)
Comment (nsoranzo) [inline test/shed_functional/base/twilltestcase.py]: Actually if you want to be consistent with `export_capsule()` method, that would be:

```
    try:
        return _commit_and_push(hgrepo, url)
    except Exception:
        return _commit_and_push(hgrepo, url)

def _commit_and_push(hgrepo, url):
    try:
        commands.push( ui.ui(), hgrepo, dest=url )
    except Exception as e:
        if str(e).find('Pushing to Tool Shed is disabled') != -1:
            return False
        else:
            raise
    return True
```

which is still not extendable to more than 2 retries, but at least avoids the code duplication.

**Comment by martenson** [comment]
refactored to reflect on @nsoranzo good idea

**Comment by nsoranzo** [comment]
:+1: Thanks @martenson!

**Comment by martenson** [comment]
please review

**Comment by martenson** [comment]
thanks @nsoranzo

**Comment by martenson** [comment]
note: we want to backport this to 16.04 if it proves working

**Pull request merged**